crates/volta-core/src/event.rs

```diff
@@ -99,7 +101,11 @@ impl EventLog {
     }
 
     pub fn add_event_start(&mut self, activity_kind: ActivityKind) {
-        self.add_event(EventKind::Start, activity_kind)
+        let argv = env::args_os()
+            .map(|arg| arg.to_string_lossy().to_string())
+            .collect::<Vec<String>>()
```

Comment:
Minor, but potentially relevant since this is on the hot path: I think we can avoid the intermediate allocation of `String` values by omitting the `to_string` call here: Map the values to `Cow<str>`, then collect those into a vector and join the vector to a string (`join` should work even when the type is `&str` or `Cow<str>`)

Reply:
I've been trying to get that to work, but I'm not sure how

```rust
let argv = env::args_os()
    .map(|arg| arg.to_string_lossy())
    .collect::<Vec<Cow<str>>>()
    .join(" ");
```

That gives me this error, and I haven't been able to work around it:

```
error[E0515]: cannot return reference to function parameter `arg`
   --> crates/volta-core/src/event.rs:108:24
    |
108 |             .map(|arg| arg.to_string_lossy())
    |                        ^^^^^^^^^^^^^^^^^^^^^ returns a reference to data owned by the current function
```

Reply:
Ohh, shoot, I missed that `args_os` iterates over _owned_ `OsString` values, so each iteration consumes the value. We may still be able to avoid the extra allocation by essentially implementing what `join` does but at the iterator level:

```rust
let argv = env::args_os()
    .enumerate()
    .fold(String::new(), |mut result, (i, arg)| {
        if i > 0 {
            result.push(' ');
        }
        result.push_str(&arg.to_string_lossy());
        result
    });
```

Reply:
I just asked about this in the Rust Discord to see if anyone there can help. Will update this discussion with anything I learn from that...

Reply:
We could also just make non-unicode arguments pay a little extra cost, where the happy path for 99% of cases should be fast:

```rust
let argv = env::args_os()
    .map(|arg| {
        arg.into_string()
            .unwrap_or_else(|s| s.to_string_lossy().into_owned())
    })
    .collect::<Vec<String>>()
    .join(" ");
```

There's also this, that someone suggested in the Rust Discord: <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c6b91f154c5b21f8288e0345dcb9c5d5> (no way do I want to go `unsafe` just to make this a little faster 🤣 )

Reply:
I like your option better though, it doesn't have the suboptimal non-unicode path that mine does.

I updated that in fa4186cf3eb5e20d92dc1d80f60f19b9c60c8bee